common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java

```diff
@@ -88,82 +93,124 @@ public OneForOneBlockFetcher(
+      if (!areMergedChunks) {
+        primaryId = Long.parseLong(blockIdParts[2]);
+      } else {
+        primaryId = Integer.parseInt(blockIdParts[2]);
+      }
+      if (!primaryIdToBlocksInfo.containsKey(primaryId)) {
+        primaryIdToBlocksInfo.put(primaryId, new BlocksInfo());
       }
-      BlocksInfo blocksInfoByMapId = mapIdToBlocksInfo.get(mapId);
-      blocksInfoByMapId.blockIds.add(blockId);
-      blocksInfoByMapId.reduceIds.add(Integer.parseInt(blockIdParts[3]));
+      BlocksInfo blocksInfoByPrimaryId = primaryIdToBlocksInfo.get(primaryId);
+      blocksInfoByPrimaryId.blockIds.add(blockId);
+      blocksInfoByPrimaryId.ids.add(Integer.parseInt(blockIdParts[3]));
       if (batchFetchEnabled) {
         // When we read continuous shuffle blocks in batch, we will reuse reduceIds in
         // FetchShuffleBlocks to store the start and end reduce id for range
         // [startReduceId, endReduceId).
         assert(blockIdParts.length == 5);
-        blocksInfoByMapId.reduceIds.add(Integer.parseInt(blockIdParts[4]));
+        blocksInfoByPrimaryId.ids.add(Integer.parseInt(blockIdParts[4]));
       }
     }
-    long[] mapIds = Longs.toArray(mapIdToBlocksInfo.keySet());
-    int[][] reduceIdArr = new int[mapIds.length][];
+    Set<Number> primaryIds = primaryIdToBlocksInfo.keySet();
+    // In case of FetchShuffleBlocks, secondaryIds are reduceIds. For FetchShuffleBlockChunks,
+    // secondaryIds are chunkIds.
+    int[][] secondaryIdsArray = new int[primaryIds.size()][];
     int blockIdIndex = 0;
-    for (int i = 0; i < mapIds.length; i++) {
-      BlocksInfo blocksInfoByMapId = mapIdToBlocksInfo.get(mapIds[i]);
-      reduceIdArr[i] = Ints.toArray(blocksInfoByMapId.reduceIds);
+    int secIndex = 0;
+    for (Number primaryId : primaryIds) {
+      BlocksInfo blocksInfoByPrimaryId = primaryIdToBlocksInfo.get(primaryId);
+      secondaryIdsArray[secIndex++] = Ints.toArray(blocksInfoByPrimaryId.ids);
 
-      // The `blockIds`'s order must be same with the read order specified in in FetchShuffleBlocks
-      // because the shuffle data's return order should match the `blockIds`'s order to ensure
-      // blockId and data match.
-      for (int j = 0; j < blocksInfoByMapId.blockIds.size(); j++) {
-        this.blockIds[blockIdIndex++] = blocksInfoByMapId.blockIds.get(j);
+      // The `blockIds`'s order must be same with the read order specified in FetchShuffleBlocks/
+      // FetchShuffleBlockChunks because the shuffle data's return order should match the
+      // `blockIds`'s order to ensure blockId and data match.
+      for (int j = 0; j < blocksInfoByPrimaryId.blockIds.size(); j++) {
+        this.blockIds[blockIdIndex++] = blocksInfoByPrimaryId.blockIds.get(j);
       }
     }
     assert(blockIdIndex == this.blockIds.length);
-
-    return new FetchShuffleBlocks(
-      appId, execId, shuffleId, mapIds, reduceIdArr, batchFetchEnabled);
+    if (!areMergedChunks) {
+      long[] mapIds = Longs.toArray(primaryIds);
```

Comment:
`Long.toArray` usage for regular block  here is existing code. In the existing code this was be done at line 141
```
long[] mapIds = Longs.toArray(mapIdToBlocksInfo.keySet());
```
I just had to move it around. Should I still change this?

I added `Ints.toArray` for shuffleChunks. The way I can make it clean/concise is by adding separate methods for converting a Set<Number> to long[] and int[]. I can add them to a utility class if it exists in this module or create a new one. Let me know what you think

Reply:
How often is this invoked ? If not that common, let us keep it as is (clarity is more important) - if it is **very** frequent, let us make it faster with a util method.

Reply:
This is invoked for each fetch request from the client. A fetch request is for multiple blocks. Since it is per fetch request, it is still frequent so I will create utils for it.